### ISSUE TITLE
Add JS, Sass, and PHP linting

### ISSUE THREAD

**Issue by fpcorso** [PopupMaker/Popup-Maker#822]
We've been wanting to set up better linting both locally as well as automated at the repo level.

This includes:

- [x] JS Linting at local level - Merged in #826 
- [x] Sass Linting at local level - Merged in #855 
- [x] PHP Linting at local level - Merged in #858 

@danieliser Has already gather a few resources for our first steps:

1. [Lint Code Like A Boss](https://webdevstudios.com/2017/04/06/lint-code-like-boss/)
2. https://www.wptide.org/local-development

A lot of what we want to accomplish can be done using [@wordpress/scripts](https://developer.wordpress.org/block-editor/packages/packages-scripts/) so that may be a good starting point.

## JS Linting
The main area for JS linting uses ESLint. There are npm modules:
https://www.npmjs.com/package/eslint
https://www.npmjs.com/package/@wordpress/eslint-plugin

After setting up an eslint config file, we can do local linting to create a command within our package.json. From there, PHPStorm also has an Eslint plugin which can be pointed at our npm modules to allow it to do linting for us.

## Sass Linting
The main packages for Sass linting are:
1. https://www.npmjs.com/package/sass-lint
2. https://www.npmjs.com/package/stylelint <-- wordpress/scripts uses this one.

If using stylelint, after setting up a `.stylelintrc.json` file, we can do local linting to create a command within our package.json as well as set up our PHPStorm with plugin to do linting for us.

## PHP Linting
The main package for PHP linting is PHPCS. We will also need to use the [WordPress Coding Standards](https://github.com/WordPress-Coding-Standards/WordPress-Coding-Standards) plugin for PHPCS.

**Comment by danieliser** [comment]
Since we are moving to using @wordpress/scripts, a lot of this is built in, so we need to work out what is & is not included. JS linting though is aleady there for sure.

**Comment by fpcorso** [comment]
@danieliser I believe @wordpress/scripts would replace our needed dependencies as well as the scripts we need to write. However, we would still need to set up PHPStorm to integrate to show the errors within the IDE. 

Additionally, if we wanted to use some of the automated repo scans, I think we still need the config files for each of the linters.

**Comment by danieliser** [comment]
@fpcorso You are correct, we can set up the .editorconfig file along with composer.json & package.json configs that should help IDE's auto enable the proper linting & tooling.

We can add supplimentary guide if there are things that can't be automated or additional PHP Storm plugins & settings they need to look at.

Then again I'm playing with VS Code and can see why people like it. So might need to write that same guide for both, which would effectively make me learn how to use it at least for our own tooling needs.

But yes, I completley agree, standardized team tooling would be ideal. 

If I can get better internet here I'd even be open to using one of the shared team code spaces built on VS Code server, then we could even co-edit and such. Any change in tooling gets populated to all teams spaces etc. Interesting concept for us to at least consider, though I love working locally 😜

**Comment by fpcorso** [comment]
@danieliser I expanded the issue description for all 3 types of linting and included parts of our conversation from here and Slack.

Have you already started some of this anywhere? If not, I can create a new branch off of `develop` to begin working on this.

**Comment by danieliser** [comment]
Yes its already there for the block functionality that went into v1.10 I believe. https://github.com/PopupMaker/Popup-Maker/blob/master/package.json#L28-L30

Now we are not currently running it in our final build processes, though I imagine to get them all passing would require massive attention off the top, then once the code is all updated to compliancy future automated tests as you write would be suitable along with IDE auto hinting and such.

**Comment by fpcorso** [comment]
@danieliser In PHPStorm, be sure to switch on automatic ESLinting if you are on `develop` or any branch off of that now.
![image](https://user-images.githubusercontent.com/9730040/84405553-49ef6c00-abd6-11ea-8261-881a7c2553fb.png)

**Comment by danieliser** [comment]
@fpcorso nice, and I see a --fix on save option too we can explore.

**Comment by fpcorso** [comment]
@danieliser Yeah, I saw that. Pretty awesome. But, I am still finding a couple of cases where I had to tweak the overrides/rules some so I am hesitant on turning on any of the auto-fixing until after we use the linting for a bit more development.

**Comment by fpcorso** [comment]
@danieliser PHPStorm also has a native integration with StyleLint. As simple as the ESLint: Just turn it on once the stylelint config file is present.
![image](https://user-images.githubusercontent.com/9730040/87831778-6bd1b380-c852-11ea-9f27-cb3fa5405cbd.png)

**Comment by fpcorso** [comment]
Moved the automated linting sub-task out into its own issue since we will wait on that until after we can go through and fix a lot of the identified warnings.

**Comment by fpcorso** [comment]
@danieliser PHP Storm has a native integration with PHPCS too. To turn on once this is merged, you can just open up Languages & Frameworks -> PHP -> Quality Tools and enabled PHPCS and point it to the one in the `vendor` folder if it doesn't automatically find it:

![image](https://user-images.githubusercontent.com/9730040/88102247-39cf9280-cb6d-11ea-972f-e63dde9e0d28.png)

From there, if you never had PHPCS enabled in PHP Storm before, you will want to go to Editor -> Inspections -> PHP -> Quality Tools to set the Coding Standard to "WordPress":

![image](https://user-images.githubusercontent.com/9730040/88102410-84e9a580-cb6d-11ea-9a78-fb31eaf2c18a.png)
